### DISCUSSION TITLE
Integrating Gephi graph visualization with Jetbrains IDEs

### DISCUSSION THREAD

**Discussion by RobertDiebels** [gephi/gephi#2781] [Q&A] [answered]
I've been looking for graph visualization libraries lately to see if they can be integrated with the Jetbrains IDEs.
Both Gephi and Jetbrains work with swing components so I believe it can be done.

Now, I have to admit I don't quite understand both the Jetbrains IDE structure and Gephi's codebase quite yet.
So, I'd like to pose a few questions to the community.

- As an expert, do you believe my assumption that integrating Gephi's swing components and the Jetbrains IDEs can be done?
- If you believe it can be done, what would the minimal set of Gephi modules be to accomplish this?
- What kind of hurdles do you foresee in an undertaking of this kind?

Thanks in advance for any help you are willing to give.

**Comment by mbastian** [comment, marked as answer]
Hi @RobertDiebels, and thanks for the inquiry.

I'm confident it can be done. You can take a look at the [Gephi Toolkit](https://github.com/gephi/gephi-toolkit), which I think would be the right way to solve this. That said, at the moment the Toolkit only provides the Preview visualization. That's the one which is easily customisable but doesn't have the performance of the Overview. But Preview module is 100% Swing and doesn't require OpenGL so there won't be integration issues because of OpenGL. The Gephi Toolkit lists the Gephi modules you need already so no need to hand-pick certain modules.

If you take a look at [this Gephi Toolkit Demo](https://github.com/gephi/gephi-toolkit-demos/tree/master/src/main/java/org/gephi/toolkit/demos/plugins/preview) you see a minimum skeleton for a Preview application, including mouse events. I would imagine this to be a good starting point.

I'm not familiar with the Jetbrains plugin system so it's hard for me to imagine potential hurdles. But from the top of my head
- Gephi requires some `-J--add-opens` flags to run on newer JREs due to some of our Netbeans dependencies. Most likely IntelliJ as well but something that would need to be checked.
- Some Gephi dependencies may clash with Jetbrains dependencies, different versions for instance.

Let us know if you run into troubles
